src/liballoc/rc.rs

```diff
@@ -571,9 +571,19 @@ impl<T: ?Sized> Rc<T> {
     /// ```
     #[stable(feature = "rc_raw", since = "1.17.0")]
     pub fn into_raw(this: Self) -> *const T {
-        let ptr: *const T = &*this;
+        let ptr: *mut RcBox<T> = NonNull::as_ptr(this.ptr);
+        let fake_ptr = ptr as *mut T;
```

Comment:
What's "fake" about this ptr?

Reply:
It's pointing to the wrong place. We create the fake pointer here so that we can take it and rewrite the pointer bit to be correct while retaining the fat pointer metadata.

Reply:
We could take the metadata directly from `ptr` though?

Reply:
Again, I'm just following the way it's done elsewhere in the file. But it's because of the type signature of `set_data_ptr`, which is `*const T -> *const T`. The cast has to happen before or after that call, and everywhere else in this file it's before.

Reply:
"we are doing undocumented magic elsewhere so we should do more of this" isn't a very convincing argument, IMO. ;)

If you don't want to change the code, please at least add some comments explaining what you just explained to me here.

Reply:
Manipulating fat pointers in this way is straight up unspecified outside the stdlib anyway, so...

I'll be completely honest in that my understanding of the fat pointer manipulation in these files is just from reading what's already done and assuming it works. I'm not 100% sure I'm correct, just closer to 90%.

I want to attack these files with a refactor, but this small fix isn't the PR to do it imho.

I guess my argument is "I'm doing it exactly like `from_raw`, just in reverse" and diverging the implementation further before a more comprehensive refactor feels wrong.

Reply:
In my last comment I just asked for you to add a comment to the code, how's that "diverging the implementation further"?

Here's a template:
```rust
        let ptr: *mut RcBox<T> = NonNull::as_ptr(this.ptr);
        mem::forget(this);
        // Obtain a ptr with the right metadata but pointing at the wrong thing.
        let fake_ptr = ptr as *mut T;

        // SAFETY: This cannot go through Deref::deref.
        // Instead, we manually offset the pointer rather than manifesting a reference.
        // This is so that the returned pointer retains the same provenance as our pointer.
        // This is required so that e.g. `get_mut` can write through the pointer
        // after the Rc is recovered through `from_raw`.
        unsafe {
            let offset = data_offset(&(*ptr).value);
            let ptr = (ptr as *mut u8).offset(offset);
            // Return `ptr` equipped with the metadata of `fake_ptr`.
            set_data_ptr(fake_ptr, ptr)
        }
```